### PR TITLE
Static sizer and timeout handlers in the pipeline

### PR DESCRIPTION
**Pull request by AlexProgrammerDE**
This simplifies all pipeline code and ensures some listeners like the sizer are always present. The code already assumed that the sizer is always there and thus causes issues. The sizer can be deactivated still now and has pretty much no performance losses from this. The profit from this PR is that there is less logic with modifying the pipeline and thus developers interacting with the channel can assume specific things about the order and placements of elements in the pipeline. This will be useful once ViaVersion is supported, and it is expected that certain elements always are in the pipeline and don't change. My plan is to also always have an encryption and compression handler in the pipeline that is controlled via AttributeKeys from netty, but for that first #828 needs to be merged. So this PR only completes the goal partially, but that's fine. PR is ready for review like it is right now.

### CODE REVIEW

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java [hunk 1]
```diff
@@ -20,6 +18,10 @@ public class BuiltinFlags {
      */
     public static final Flag<Boolean> TCP_FAST_OPEN = new Flag<>("tcp-fast-open", Boolean.class);
 
+    public static final Flag<Integer> CLIENT_CONNECT_TIMEOUT = new Flag<>("client-connect-timeout", Integer.class);
+    public static final Flag<Integer> READ_TIMEOUT = new Flag<>("read-timeout", Integer.class);
+    public static final Flag<Integer> WRITE_TIMEOUT = new Flag<>("write-timeout", Integer.class);
```
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java]: javadoc that these are in seconds, and that they may be zero or negative for the timeout checks to be disabled

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java [hunk 2]
```diff
@@ -6,8 +6,6 @@
  * Built-in PacketLib session flags.
  */
 public class BuiltinFlags {
-    public static final Flag<Boolean> ENABLE_CLIENT_PROXY_PROTOCOL = new Flag<>("enable-client-proxy-protocol", Boolean.class);
-
     public static final Flag<InetSocketAddress> CLIENT_PROXIED_ADDRESS = new Flag<>("client-proxied-address", InetSocketAddress.class);
```
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java:14]: javadoc that null == support disabled

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java [hunk 3]
```diff
@@ -206,54 +201,52 @@ private InetSocketAddress resolveAddress() {
...
+                    pipeline.addLast("proxy", new Socks5ProxyHandler(proxy.address()));
                 }
-                default -> throw new UnsupportedOperationException("Unsupported proxy type: " + proxy.type());
             }
+            default -> throw new UnsupportedOperationException("Unsupported proxy type: " + proxy.type());
         }
     }
 
-    private void addHAProxySupport(ChannelPipeline pipeline) {
+    private void addHAProxySupport(Channel channel) {
```
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java]: i feel like now might be a good time to rename it `initializeHAProxySupport` considering it adds the encoder, sends the message, then removes it.

**Review by Konicai — CHANGES_REQUESTED** [review]
looking good

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java [hunk 4]
```diff
@@ -20,6 +22,22 @@ public class BuiltinFlags {
...
+     * Read timeout in seconds.
+     * Used by both server and client.
+     */
+    public static final Flag<Integer> READ_TIMEOUT = new Flag<>("read-timeout", Integer.class);
+    /**
+     * Write timeout in seconds.
+     * Used by both server and client.
+     */
+    public static final Flag<Integer> WRITE_TIMEOUT = new Flag<>("write-timeout", Integer.class);
+
```
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java:40]:
```suggestion
    /**
     * Connection timeout in seconds.
     * Only used by the client.
     */
    public static final Flag<Integer> CLIENT_CONNECT_TIMEOUT = new Flag<>("client-connect-timeout", Integer.class);
    
    /**
     * Read timeout in seconds.
     * Used by both server and client.
     */
    public static final Flag<Integer> READ_TIMEOUT = new Flag<>("read-timeout", Integer.class);
    
    /**
     * Write timeout in seconds.
     * Used by both server and client.
     */
    public static final Flag<Integer> WRITE_TIMEOUT = new Flag<>("write-timeout", Integer.class);

```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java [hunk 5]
```diff
@@ -20,6 +22,22 @@ public class BuiltinFlags {
...
+     * Only used by the client.
+     */
+    public static final Flag<Integer> CLIENT_CONNECT_TIMEOUT = new Flag<>("client-connect-timeout", Integer.class);
+    /**
+     * Read timeout in seconds.
+     * Used by both server and client.
+     */
+    public static final Flag<Integer> READ_TIMEOUT = new Flag<>("read-timeout", Integer.class);
+    /**
+     * Write timeout in seconds.
```
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java]:
```suggestion
    public static final Flag<Integer> CLIENT_CONNECT_TIMEOUT = new Flag<>("client-connect-timeout", Integer.class);
    
    /**
     * Read timeout in seconds.
     * Used by both server and client.
     */
    public static final Flag<Integer> READ_TIMEOUT = new Flag<>("read-timeout", Integer.class);
    
    /**
     * Write timeout in seconds.
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java [hunk 6]
```diff
@@ -6,8 +6,10 @@
  * Built-in PacketLib session flags.
  */
 public class BuiltinFlags {
-    public static final Flag<Boolean> ENABLE_CLIENT_PROXY_PROTOCOL = new Flag<>("enable-client-proxy-protocol", Boolean.class);
-
+    /**
+     * Enables HAProxy protocol support.
```
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java:11]:
```suggestion

    /**
     * Enables HAProxy protocol support.
```

**Review by Konicai — APPROVED** [review]

**Review by Redned235 — APPROVED** [review]

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java [hunk 7]
```diff
@@ -6,8 +6,11 @@
  * Built-in PacketLib session flags.
  */
 public class BuiltinFlags {
-    public static final Flag<Boolean> ENABLE_CLIENT_PROXY_PROTOCOL = new Flag<>("enable-client-proxy-protocol", Boolean.class);
 
+    /**
+     * Enables HAProxy protocol support.
+     * When this value is not null it means the ip and port the client claims the connection is from.
```
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java]:
```suggestion
     * When this value is not null it represents the ip and port the client claims the connection is from.
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java [hunk 8]
```diff
@@ -20,6 +23,24 @@ public class BuiltinFlags {
...
 
+    /**
+     * Connection timeout in seconds.
+     * Only used by the client.
+     */
+    public static final Flag<Integer> CLIENT_CONNECT_TIMEOUT = new Flag<>("client-connect-timeout", Integer.class);
+    
+    /**
+     * Read timeout in seconds.
+     * Used by both server and client.
```
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java]:
```suggestion
     * Used by both the server and client.
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java [hunk 9]
```diff
@@ -20,6 +23,24 @@ public class BuiltinFlags {
...
+     */
+    public static final Flag<Integer> READ_TIMEOUT = new Flag<>("read-timeout", Integer.class);
+    
+    /**
+     * Write timeout in seconds.
+     * Used by both server and client.
```
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/BuiltinFlags.java]:
```suggestion
     * Used by both the server and client.
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java [hunk 10]
```diff
@@ -77,75 +75,62 @@ public void initChannel(Channel channel) {
...
         if (this.channel != null) {
             if (this.channel.isOpen()) {
-                ChannelFuture future = this.channel.close();
-                if (wait) {
-                    try {
-                        future.sync();
-                    } catch (InterruptedException e) {
+                CompletableFuture<Void> handleFuture = new CompletableFuture<>();
+                this.channel.close().addListener((ChannelFutureListener) future1 -> {
+                    if (future1.isSuccess()) {
```
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java]:
```suggestion
                this.channel.close().addListener((ChannelFutureListener) future -> {
                    if (future.isSuccess()) {
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java [hunk 11]
```diff
@@ -77,75 +75,62 @@ public void initChannel(Channel channel) {
...
-            } catch (InterruptedException e) {
+        CompletableFuture<Void> handleFuture = new CompletableFuture<>();
+        bootstrap.bind().addListener((ChannelFutureListener) future1 -> {
+            if (future1.isSuccess()) {
+                channel = future1.channel();
+                if (callback != null) {
+                    callback.run();
+                }
+            } else {
+                log.error("Failed to bind connection listener.", future1.cause());
```
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java]:
```suggestion
        bootstrap.bind().addListener((ChannelFutureListener) future -> {
            if (future.isSuccess()) {
                channel = future.channel();
                if (callback != null) {
                    callback.run();
                }
            } else {
                log.error("Failed to bind connection listener.", future.cause());
```
There's no other variable called future; so this works too (and looks a bit cleaner)

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java [hunk 12]
```diff
@@ -77,75 +75,62 @@ public void initChannel(Channel channel) {
...
-                        future.sync();
-                    } catch (InterruptedException e) {
+                CompletableFuture<Void> handleFuture = new CompletableFuture<>();
+                this.channel.close().addListener((ChannelFutureListener) future1 -> {
+                    if (future1.isSuccess()) {
+                        if (callback != null) {
+                            callback.run();
+                        }
+                    } else {
+                        log.error("Failed to close connection listener.", future1.cause());
```
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java]:
```suggestion
                        log.error("Failed to close connection listener.", future.cause());
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java [hunk 13]
```diff
@@ -77,75 +75,62 @@ public void initChannel(Channel channel) {
...
         if (this.group != null) {
-            Future<?> future = this.group.shutdownGracefully();
-            if (wait) {
-                try {
-                    future.sync();
-                } catch (InterruptedException e) {
+            CompletableFuture<Void> handleFuture = new CompletableFuture<>();
+            this.group.shutdownGracefully().addListener(future1 -> {
+                if (!future1.isSuccess()) {
+                    log.debug("Failed to close connection listener.", future1.cause());
```
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java]:
```suggestion
            this.group.shutdownGracefully().addListener(future -> {
                if (!future.isSuccess()) {
                    log.debug("Failed to close connection listener.", future.cause());
```

**Review by onebeastchris — APPROVED** [review]
last nitpicks, let's merge after these

**Pull request merged**